File: Source/WebKit2/UIProcess/sling/WebViewContext.cpp
/*
 * Copyright (C) 2015 Naver Corp. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include "WebViewContext.h"

#include "SWKView.h"
#include "WebEditableContext.h"
#include "WebRenderContext.h"

namespace webkit {

std::shared_ptr<WebViewContext> WebViewContext::create(std::shared_ptr<SWKView> view)
{
    return std::shared_ptr<WebViewContext>(new WebViewContext(view));
}

WebViewContext::WebViewContext(std::shared_ptr<SWKView> view)
    : m_view(view)
{
}

WebViewContext::~WebViewContext()
{
}

void WebViewContext::attached(view::ContentView& view)
{
    m_view->onAttachedToView(view);
}

void WebViewContext::detached()
{
    m_view->onDetachedFromView();
}

void WebViewContext::setVisible(bool visible)
{
    m_view->onVisibilityChanged(visible);
}

void WebViewContext::setFocus(bool focus)
{
    m_view->onFocusChanged(focus);
}

void WebViewContext::setActive(bool active)
{
    m_view->setActive(active);
}

bool WebViewContext::isVisible()
{
    return m_view->isVisible();
}

bool WebViewContext::isFocused()
{
    return m_view->isFocused();
}

bool WebViewContext::isActive()
{
    return m_view->isActive();
}

void WebViewContext::layout(int32_t width, int32_t height)
{
    m_view->onLayout(width, height);
}

void WebViewContext::displayLocation(int32_t x, int32_t y)
{
    m_view->onDisplayLocation(x, y);
}

void WebViewContext::onConfigurationChanged(const view::Configuration& config)
{
    m_view->onConfigurationChanged(config);
}

void WebViewContext::pause()
{
    m_view->onPause();
}

void WebViewContext::resume()
{
    m_view->onResume();
}

view::MotionDetector::Settings WebViewContext::motionSettings() const
{
    return m_view->getMotionSettings();
}

void WebViewContext::sendSingleTap(float x, float y)
{
    m_view->sendSingleTap(x, y);
}

void WebViewContext::sendDoubleTap(float x, float y)
{
    m_view->sendDoubleTap(x, y);
}

bool WebViewContext::sendGenericEvent(const view::UIEvent& event)
{
    return m_view->sendGenericEvent(event);
}

std::shared_ptr<view::EditableContext> WebViewContext::editable()
{
    return WebEditableContext::create(m_view);
}

int32_t WebViewContext::width() const
{
    return m_view->getWidth();
}

int32_t WebViewContext::height() const
{
    return m_view->getHeight();
}

int32_t WebViewContext::contentWidth() const
{
    return m_view->getContentWidth();
}

int32_t WebViewContext::contentHeight() const
{
    return m_view->getContentHeight();
}

float WebViewContext::currentScale() const
{
    return m_view->getScale();
}

float WebViewContext::minimumScale() const
{
    return m_view->getMinimumScale();
}

float WebViewContext::maximumScale() const
{
    return m_view->getMaximumScale();
}

void WebViewContext::scrollBegin(float x, float y)
{
    return m_view->scrollBegin(x, y);
}

bool WebViewContext::scrollBy(float deltaX, float deltaY)
{
    return m_view->scrollBy(deltaX, deltaY, true);
}

bool WebViewContext::scrollTo(float x, float y)
{
    return false;
}

void WebViewContext::scrollEnd(float x, float y)
{
    return m_view->scrollEnd(x, y);
}

void WebViewContext::pinchBegin(float x, float y)
{
    m_view->pinchBegin(x, y);
}

void WebViewContext::pinchTo(const view::RectF& rect)
{
    m_view->pinchTo(rect);
}

void WebViewContext::pinchBy(float x, float y, float deltaScale)
{
    m_view->pinchBy(x, y, deltaScale);
}

void WebViewContext::pinchEnd()
{
    m_view->pinchEnd();
}

view::RectF WebViewContext::viewport() const
{
    return m_view->getViewport();
}

void WebViewContext::invalidate()
{
    m_view->invalidate();
}

void WebViewContext::invalidate(const view::Rect& rect)
{
    m_view->invalidate(rect);
}

std::shared_ptr<view::RenderContext> WebViewContext::renderer()
{
    if (!m_renderer)
        m_renderer = WebRenderContext::create(m_view);

    return m_renderer;
}

} // namespace webkit

File: Source/WebKit2/UIProcess/sling/WebViewContext.java
package labs.naver.webkit.view;

import android.annotation.SuppressLint;
import android.content.res.Configuration;
import android.graphics.Rect;
import android.graphics.RectF;
import android.view.View;

import labs.naver.view.ContentView;
import labs.naver.view.EditableContext;
import labs.naver.view.MotionDetector.Settings;
import labs.naver.view.RenderContext;
import labs.naver.view.UIEvent;
import labs.naver.view.ViewContext;
import labs.naver.webkit.api.SWKView;

public class WebViewContext implements ViewContext {

    @Override
    public void attached(ContentView contentView) {
        mView.onAttachedToView(contentView);
    }

    @Override
    public void detached() {
        mView.onDetachedFromView();
    }

    @Override
    public void setVisible(int visibility) {
        mView.onVisibilityChanged(visibility == View.VISIBLE);
    }

    @Override
    public void setFocus(boolean focus) {
        mView.onFocusChanged(focus);
    }

    @Override
    public void setActive(boolean active) {
        mView.setActive(active);
    }

    @Override
    public boolean isVisible() {
        return mView.isVisible();
    }

    @Override
    public boolean isFocused() {
        return mView.isFocused();
    }

    @Override
    public boolean isActive() {
        return mView.isActive();
    }

    @SuppressLint("WrongCall")
    @Override
    public void layout(int width, int height) {
        mView.onLayout(width, height);
    }

    @Override
    public void onConfigurationChanged(Configuration config) {
        mView.onConfigurationChanged(config);
    }

    @Override
    public void pause() {
        mView.onPause();
    }

    @Override
    public void resume() {
        mView.onResume();
    }

    @Override
    public Settings motionSettings() {
        return mView.getMotionSettings();
    }

    @Override
    public void sendSingleTap(float x, float y) {
        mView.sendSingleTap(x, y);
    }

    @Override
    public void sendDoubleTap(float x, float y) {
        mView.sendDoubleTap(x, y);
    }

    @Override
    public boolean sendGenericEvent(UIEvent event) {
        return mView.sendGenericEvent(event);
    }

    @Override
    public EditableContext editable() {
        return new WebEditableContext(mView);
    }

    @Override
    public int width() {
        return mView.getWidth();
    }

    @Override
    public int height() {
        return mView.getHeight();
    }

    @Override
    public int contentWidth() {
        return mView.getContentWidth();
    }

    @Override
    public int contentHeight() {
        return mView.getContentHeight();
    }

    @Override
    public float currentScale() {
        return mView.getScale();
    }

    @Override
    public float minimumScale() {
        return mView.getMinimumScale();
    }

    @Override
    public float maximumScale() {
        return mView.getMaximumScale();
    }

    @Override
    public void scrollBegin(float x, float y) {
        mView.scrollBegin(x, y);
    }

    @Override
    public boolean scrollBy(float deltaX, float deltaY) {
        return mView.scrollBy(deltaX, deltaY, true);
    }

    @Override
    public boolean scrollTo(float x, float y) {
        return mView.scrollTo(x, y);
    }

    @Override
    public void scrollEnd(float x, float y) {
        mView.scrollEnd(x, y);
    }

    @Override
    public void pinchBegin(float x, float y) {
        mView.pinchBegin(x, y);
    }

    @Override
    public void pinchTo(RectF rect) {
        mView.pinchTo(rect);
    }

    @Override
    public void pinchBy(float x, float y, float deltaScale) {
        mView.pinchBy(x, y, deltaScale);
    }

    @Override
    public void pinchEnd() {
        mView.pinchEnd();
    }

    @Override
    public RectF viewport() {
        return mView.getViewport();
    }

    @Override
    public void invalidate() {
        mView.invalidate();
    }

    @Override
    public void invalidate(Rect rect) {
        mView.invalidate(rect);
    }

    @Override
    public RenderContext renderer() {
        if (mRenderer == null)
            mRenderer = new WebRenderContext(mView);

        return mRenderer;
    }

    private final SWKView mView;
    private RenderContext mRenderer;

    public WebViewContext(SWKView view) {
        mView = view;
    }
}
